lib/nostr.js

```diff
@@ -15,6 +17,149 @@ export const DEFAULT_CROSSPOSTING_RELAYS = [
   'wss://relay.mutinywallet.com/'
 ]
 
+export class Relay {
+  constructor (relayUrl) {
+    const ws = new WebSocket(relayUrl)
+
+    ws.onmessage = function (msg) {
+      const [type, notice] = JSON.parse(msg.data)
+      if (type === 'NOTICE') {
+        console.log('relay notice:', notice)
+      }
+    }
+
+    ws.onerror = function (err) {
+      console.error('websocket error: ' + err)
+      this.error = err
+    }
+
+    this.ws = ws
+  }
+
+  static async connect (url, { timeout } = {}) {
+    const relay = new Relay(url)
+    await relay.waitUntilConnected({ timeout })
+    return relay
+  }
+
+  get connected () {
+    return this.ws.readyState === WebSocket.OPEN
+  }
+
+  get closed () {
+    return this.ws.readyState === WebSocket.CLOSING || this.ws.readyState === WebSocket.CLOSED
+  }
+
+  async waitUntilConnected ({ timeout } = {}) {
+    let interval
+
+    const checkPromise = new Promise((resolve, reject) => {
+      interval = setInterval(() => {
+        if (this.connected) {
+          resolve()
+        }
+        if (this.closed) {
+          reject(new Error(`failed to connect to ${this.url}: ` + this.error))
+        }
+      }, 100)
+    })
+
+    try {
+      return await withTimeout(checkPromise, timeout)
+    } finally {
+      clearInterval(interval)
+    }
+  }
+
+  close () {
+    const state = this.ws.readyState
+    if (state !== WebSocket.CLOSING && state !== WebSocket.CLOSED) {
+      this.ws.close()
+    }
+  }
+
+  async publish (event, { timeout } = {}) {
+    const ws = this.ws
+
+    let listener
+    const ackPromise = new Promise((resolve, reject) => {
+      ws.send(JSON.stringify(['EVENT', event]))
+
+      listener = function onmessage (msg) {
+        const [type, eventId, accepted, reason] = JSON.parse(msg.data)
+
+        if (type !== 'OK' || eventId !== event.id) return
+
+        if (accepted) {
+          resolve(eventId)
+        } else {
+          reject(new Error(reason || `event rejected: ${eventId}`))
+        }
+      }
+
+      ws.addEventListener('message', listener)
+    })
+
+    try {
+      return await withTimeout(ackPromise, timeout)
+    } finally {
+      ws.removeEventListener('message', listener)
+    }
+  }
+
+  async fetch (filter, { timeout } = {}) {
+    const ws = this.ws
+
+    let listener
+    const ackPromise = new Promise((resolve, reject) => {
+      const id = crypto.randomUUID()
```

Comment:
Mhh, NWC autowithdrawals (#1310) fail because of this error:

```
worker        | ReferenceError: crypto is not defined
worker        |     at Object.encrypt (file:///app/node_modules/nostr-tools/lib/esm/index.js:1282:19)
worker        |     at nwcCall (file:///app/wallets/nwc/index.js:47:35)
worker        |     at async createInvoice (file:///app/wallets/nwc/server.js:13:18)
worker        |     at async createInvoice (file:///app/wallets/server.js:48:23)
worker        |     at async autoWithdraw (file:///app/worker/autowithdraw.js:41:31)
worker        |     at async file:///app/worker/index.js:80:9
worker        |     at resolveWithinSeconds (/app/node_modules/pg-boss/src/manager.js:35:14)
worker        |     at <anonymous> (/app/node_modules/p-map/index.js:57:22)
```

`import crypto from 'crypto'` did not help

Reply:
Fixed in 5515c63c